File: player/lib/core/compatibility/compatibility.dart
/// Version floors this player declares about the servers it connects to.
///
/// The mirror of these lives server-side in `lib/mydia/compatibility.ex`, which
/// declares the oldest *player* a server works with. The two are different
/// facts and neither derives from the other: a server cannot know what a future
/// player needs, and a player cannot know what a future server drops.
///
/// This direction matters more often in practice. Mobile and desktop players
/// update themselves from GitHub releases while a self-hosted server stays on
/// whatever the operator last pulled, so "player newer than server" is the
/// common mismatch, and an old server has no way to warn about it.
///
/// ## When to bump
///
/// Bump [minServerVersion] to the release you are shipping when the player
/// starts depending on something older servers do not provide: a GraphQL field
/// added in that release, a changed streaming contract, a new endpoint.
///
/// Bump [recommendedServerVersion] when an older server still works but the
/// experience is degraded.
///
/// Do not bump either for routine releases. A floor that moves every release
/// trains operators to ignore the banner.
///
/// ## Last changed
///
/// - 0.9.0: initial baseline, set to the release this mechanism shipped in.
class Compatibility {
  const Compatibility._();

  /// The oldest server version this player works with.
  static const minServerVersion = '0.9.0';

  /// The oldest server version this player would rather talk to.
  static const recommendedServerVersion = '0.9.0';
}

File: lib/mydia/compatibility.ex
defmodule Mydia.Compatibility do
  @moduledoc """
  Version floors this server declares to connecting players.

  Mydia servers and players ship from the same release tag but operators
  upgrade on their own timeline, so both sides can be on different releases at
  once. These constants let a server tell a player that it is too old to work
  correctly, instead of the mismatch surfacing as an unexplained failure.

  ## When to bump

  Bump `@min_player_version` to the release you are shipping when that release
  makes a change a older player cannot cope with: removing or renaming a
  GraphQL field the player selects, changing the meaning of a response, or
  changing a streaming or download contract.

  Bump `@recommended_player_version` when a release adds something a player
  needs to take advantage of, but whose absence degrades gracefully.

  Do not bump either for routine releases. Every bump nags every operator
  running an older player, and a floor that moves on every release trains
  people to ignore the banner.

  The player carries the mirror of these values, the oldest server it works
  with, in `player/lib/core/compatibility/compatibility.dart`. The two are
  different facts and neither derives from the other.

  ## Last changed

  - 0.9.0: initial baseline, set to the release this mechanism shipped in.
  """

  @min_player_version "0.9.0"
  @recommended_player_version "0.9.0"

  @doc "The oldest player version this server works with."
  @spec min_player_version() :: String.t()
  def min_player_version, do: @min_player_version

  @doc "The oldest player version this server would rather you were running."
  @spec recommended_player_version() :: String.t()
  def recommended_player_version, do: @recommended_player_version
end
